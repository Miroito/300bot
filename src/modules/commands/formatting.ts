import { Character, CharacterWithOutfitWithLeader } from '@planetside/types'

const factionEmojis = [
  'Unknown Faction',
  '{emoji:faction_logo_vs|VS}',
  '{emoji:faction_logo_nc|NC}',
  '{emoji:faction_logo_tr|TR}',
  '{emoji:faction_logo_ns|NS}',
]

export const formatChacarcterWithFaction = (
  character: Character | CharacterWithOutfitWithLeader,
): string => {
  return `**${character.name.first}** ${
    factionEmojis[Number(character.factionId)]
  }${
    character.factionId === '4' &&
    'outfitMember' in character &&
    character.outfitMember
      ? ` ${
          factionEmojis[Number(character.outfitMember.outfit.leader.factionId)]
        }`
      : ''
  }`
}
