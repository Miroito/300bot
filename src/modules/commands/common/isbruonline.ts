import { sentence } from '@app/utils/language'
import { Command } from '@commands/CommandHandler'
import { formatChacarcterWithFaction } from '@commands/formatting'
import { bruCharactersDatabase } from '@database/brucharacters'
import { censusApi } from '@planetside/CensusApi'

export default new Command({
  keyword: 'isbruonline',
  alias: ['isbruoffline'],
  description: 'check if Bru is online',
  help: 'Usage: `{prefix}isbruonline` - checks if Bru is online',
  callback: async ({ args, alias, reply }) => {
    if (args.length > 0) return

    const statuses =
      await censusApi.getCharactersWithOutfitLeaderAndOnlineStatus({
        characterId: bruCharactersDatabase.keys,
      })

    if (!statuses) {
      return reply('Bru has no existing characters.')
    }

    const formattedCharacters = statuses
      .filter((character) => character.onlineStatus !== '0')
      .map(formatChacarcterWithFaction)

    const online = formattedCharacters.length > 0
    if (alias === 'isbruonline') {
      return reply(
        online
          ? `Yes, Bru is online as ${sentence(formattedCharacters)}!`
          : 'No, Bru is offline.',
      )
    } else {
      return reply(
        online
          ? `No, Bru is online as ${sentence(formattedCharacters)}.`
          : 'Yes, Bru is offline!',
      )
    }
  },
})
