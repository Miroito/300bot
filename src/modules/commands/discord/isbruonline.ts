import { sentence } from '@app/utils/language'
import { Command } from '@commands/CommandHandler'
import { DiscordParams } from '@commands/params'
import { validateArgumentNumber } from '@commands/validators'
import { bruCharactersDatabase } from '@database/brucharacters'
import { displayCharacter } from '@discord/resources'
import { censusApi } from '@planetside/CensusApi'
import { TestDiscordParams } from '@test/utils/commands'

export default new Command<DiscordParams | TestDiscordParams>({
  keyword: 'isbruonline',
  alias: ['isbruoffline'],
  description: 'check if Bru is online',
  help: 'Usage: `{prefix}isbruonline` - checks if Bru is online',
  callback: async ({ args, alias, reply, env }) => {
    validateArgumentNumber(args.length, 0)

    const statuses =
      await censusApi.getCharacterOutfitLeaderFactionAndOnlineStatus({
        characterId: bruCharactersDatabase.keys.join(','),
      })

    if (!statuses) {
      return reply(
        'Bru has no existing characters, please verify using the brucharacters command',
      )
    }
    const channelOpt = env.message ? env.message.channel : null

    const characterList = statuses
      .filter((c) => c.onlineStatus === '10')
      .map((c) => `${displayCharacter(c, channelOpt)}`)

    const online = characterList.length > 0
    if (alias === 'isbruonline') {
      return reply(
        online
          ? `Yes, Bru is online as ${sentence(characterList)}!`
          : 'No, Bru is offline.',
      )
    } else {
      return reply(
        online
          ? `No, Bru is online as ${sentence(characterList)}.`
          : 'Yes, Bru is offline!',
      )
    }
  },
})
