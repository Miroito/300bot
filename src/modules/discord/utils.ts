import discord from 'discord.js'
import { schedule } from '@app/utils/async'
import { log } from '@app/utils/log'
import { reviveDatabase } from '@database/revive'

export const getTextChannel = (
  client: discord.Client,
  id: string,
): discord.TextChannel | null => {
  const channel = client.channels.resolve(id)
  if (channel instanceof discord.TextChannel) {
    return channel
  }
  return null
}

export const getDMChannel = async (
  client: discord.Client,
  id: string,
): Promise<discord.DMChannel | null> => {
  const channel = await client.channels.fetch(id)
  if (channel instanceof discord.DMChannel) {
    return channel
  }
  return null
}

export const formatWithEmojis = (
  channel: discord.Channel,
  message: string,
): string => {
  // eslint-disable-next-line unicorn/no-unsafe-regex
  const emojiExpressionRegex = /{emoji:([^|]*)(?:\|([^}]*))?}/g

  if (channel instanceof discord.TextChannel) {
    return message.replace(
      emojiExpressionRegex,
      (_match, emojiName, fallbackText) => {
        const emoji = channel.guild.emojis.cache.find(
          ({ name }) => name === emojiName,
        )
        if (emoji) {
          return emoji.toString()
        } else if (typeof fallbackText === 'string') {
          return fallbackText
        } else {
          return ''
        }
      },
    )
  } else {
    return message.replace(
      emojiExpressionRegex,
      (_match, _emojiName, fallbackText) => {
        if (typeof fallbackText === 'string') {
          return fallbackText
        } else {
          return ''
        }
      },
    )
  }
}

export const removeReaction = async (
  reaction: discord.MessageReaction,
  user: discord.User,
): Promise<void> => {
  try {
    await reaction.users.remove(user)
  } catch (error) {
    log.error('Error removing reaction:', error)
  }
}

export const reviveMember = async (
  member: discord.GuildMember,
  deadRole: discord.Role,
): Promise<void> => {
  const guildId = member.guild.id
  const userId = member.id

  await member.roles.remove(deadRole)
  await reviveDatabase.delete(`${guildId}.${userId}`)
}

export const killMember = async (
  member: discord.GuildMember,
  deadRole: discord.Role,
  reviveDelayMs: number,
): Promise<void> => {
  const userId = member.id
  const guildId = member.guild.id

  const reviveTime = Date.now() + reviveDelayMs

  await reviveDatabase.set(`${guildId}.${userId}`, reviveTime)
  await member.roles.add(deadRole)

  void schedule(async () => {
    try {
      await reviveMember(member, deadRole)
    } catch (e) {
      log.error('Cannot revive member:', e)
    }
  }, reviveDelayMs)
}
