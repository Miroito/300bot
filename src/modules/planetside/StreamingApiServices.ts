import { getUTCShort } from '@app/utils/time'
import { dmTrackerDatabase } from '@database/dmtracker'
import { client } from '@discord/client'
import { formatWithEmojis, getDMChannel, getTextChannel } from '@discord/utils'
import { censusApi } from '@planetside/CensusApi'
import { streamingApi } from '@planetside/StreamingApi'
import {
  getAlertType,
  getAlertState,
  getContinentName,
} from '@planetside/resources'
import { alertTrackerDatabase } from '../database/alerttracker'

export const startTrackingService = (): void => {
  const sendCharacterStatus = async (characterId: string, online: boolean) => {
    const channelIds = dmTrackerDatabase.get(characterId)
    if (!channelIds) return
    const character = await censusApi.getCharacterName({ characterId })
    if (!character) return
    const characterName = character.name.first
    const message = online
      ? `**${characterName}** is online!`
      : `**${characterName}** is offline.`
    Object.keys(channelIds).forEach((channelId) => {
      void (async () => {
        const channel = await getDMChannel(client, channelId)
        if (channel) {
          void channel.send(message)
        }
      })()
    })
  }

  streamingApi.on('playerLogin', async ({ characterId }) => {
    await sendCharacterStatus(characterId, true)
  })
  streamingApi.on('playerLogout', async ({ characterId }) => {
    await sendCharacterStatus(characterId, false)
  })
}

export const startAlertTrackerService = (): void => {
  const sendAlertMessage = (message: string) => {
    message = `[${getUTCShort()}] ${message}`
    const channelIds = Object.values(alertTrackerDatabase.root)
    channelIds.forEach((channelId) => {
      const channel = getTextChannel(client, channelId)
      if (channel) {
        void channel.send(formatWithEmojis(channel, message))
      }
    })
  }

  streamingApi.on('continentLock', ({ zoneId }) => {
    const continentName = getContinentName(zoneId)
    if (!continentName) return
    sendAlertMessage(`:lock: **${continentName}** has been locked.`)
  })
  streamingApi.on('continentUnlock', ({ zoneId }) => {
    const continentName = getContinentName(zoneId)
    if (!continentName) return
    sendAlertMessage(`:unlock: **${continentName}** has been unlocked.`)
  })
  streamingApi.on('metagameEvent', (event) => {
    const alertType = getAlertType(event.metagameEventId)
    const alertState = getAlertState(event.metagameEventState)
    if (!alertType || !alertState) return
    const [continentName, factionAlias] = alertType

    const factionData = [
      {
        alias: 'TR',
        emoji: '{emoji:faction_logo_tr}',
        emojiWithFallback: '{emoji:faction_logo_tr|**TR**}',
        result: Math.floor(Number(event.factionTr)),
      },
      {
        alias: 'NC',
        emoji: '{emoji:faction_logo_nc}',
        emojiWithFallback: '{emoji:faction_logo_nc|**NC**}',
        result: Math.floor(Number(event.factionNc)),
      },
      {
        alias: 'VS',
        emoji: '{emoji:faction_logo_vs}',
        emojiWithFallback: '{emoji:faction_logo_vs|**VS**}',
        result: Math.floor(Number(event.factionVs)),
      },
    ]

    if (alertState === 'started') {
      const faction = factionData.find((item) => item.alias === factionAlias)
      if (!faction) return
      sendAlertMessage(
        `${faction.emoji} **${faction.alias}** have started an alert on **${continentName}** with **${faction.result}%** territory control.`,
      )
    } else {
      const factionResults = factionData
        .map(
          ({ emojiWithFallback, result }) =>
            `${emojiWithFallback} **${result}%**`,
        )
        .join(' ')

      sendAlertMessage(
        `:rotating_light: The alert on **${continentName}** has ended: ${factionResults}`,
      )
    }
  })
}
