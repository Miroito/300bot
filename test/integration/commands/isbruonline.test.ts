import isbruonline from '@commands/discord/isbruonline'
import { getCommandRunner } from '@test/utils/commands'
import { mockCensusApi } from '@test/utils/planetside'

describe('isbruonline', () => {
  const runCommand = getCommandRunner(isbruonline)

  it('should say Bru is online and list the character(s) he is online with', async () => {
    const requestData = mockCensusApi({
      character_list: [
        {
          character_id: 'testlist',
          name: { first: 'MiroitoVS', first_lower: 'miroitovs' },
          faction_id: '1',
          head_id: '5',
          title_id: '0',
          times: {
            creation: '1502020291',
            creation_date: '2017-08-06 11:51:31.0',
            last_save: '1659281631',
            last_save_date: '2022-07-31 15:33:51.0',
            last_login: '1659269547',
            last_login_date: '2022-07-31 12:12:27.0',
            login_count: '305',
            minutes_played: '20631',
          },
          certs: {
            earned_points: '63245',
            gifted_points: '8742',
            spent_points: '62563',
            available_points: '9424',
            percent_to_next: '0.888',
          },
          battle_rank: { percent_to_next: '35', value: '88' },
          profile_id: '20',
          daily_ribbon: {
            count: '0',
            time: '1659225600',
            date: '2022-07-31 00:00:00.0',
          },
          prestige_level: '0',
          online_status: '10',
          outfit_member: {
            outfit_id: '37512545478660293',
            outfit: {
              leader_character_id: '5428011263355078529',
              leader: { faction_id: '1' },
            },
          },
        },
        {
          character_id: 'testlist',
          name: { first: 'Number5Johnny', first_lower: 'number5johnny' },
          faction_id: '4',
          head_id: '5',
          title_id: '0',
          times: {
            creation: '1641022496',
            creation_date: '2022-01-01 07:34:56.0',
            last_save: '1675632997',
            last_save_date: '2023-02-05 21:36:37.0',
            last_login: '1675612529',
            last_login_date: '2023-02-05 15:55:29.0',
            login_count: '229',
            minutes_played: '36020',
          },
          certs: {
            earned_points: '311191',
            gifted_points: '11132',
            spent_points: '300806',
            available_points: '21517',
            percent_to_next: '0.1351111111109',
          },
          battle_rank: { percent_to_next: '54', value: '61' },
          profile_id: '190',
          daily_ribbon: {
            count: '0',
            time: '1675555200',
            date: '2023-02-05 00:00:00.0',
          },
          prestige_level: '3',
          online_status: '10',
          outfit_member: {
            outfit_id: '37588148218236901',
            outfit: {
              leader_character_id: '5429269171559319377',
              leader: { faction_id: '3' },
            },
          },
        },
      ],
      returned: 2,
    })
    const reply = await runCommand('+isbruonline')
    expect(requestData).toMatchSnapshot()
    expect(reply).toEqual(
      'Yes, Bru is online as **MiroitoVS** VS and **Number5Johnny** NS TR!',
    )
  })

  it('should say Bru is offline', async () => {
    const requestData = mockCensusApi({
      character_list: [
        {
          character_id: '5428631729584582401',
          name: { first: 'MiroitoVS', first_lower: 'miroitovs' },
          faction_id: '1',
          head_id: '5',
          title_id: '0',
          times: {
            creation: '1502020291',
            creation_date: '2017-08-06 11:51:31.0',
            last_save: '1659281631',
            last_save_date: '2022-07-31 15:33:51.0',
            last_login: '1659269547',
            last_login_date: '2022-07-31 12:12:27.0',
            login_count: '305',
            minutes_played: '20631',
          },
          certs: {
            earned_points: '63245',
            gifted_points: '8742',
            spent_points: '62563',
            available_points: '9424',
            percent_to_next: '0.888',
          },
          battle_rank: { percent_to_next: '35', value: '88' },
          profile_id: '20',
          daily_ribbon: {
            count: '0',
            time: '1659225600',
            date: '2022-07-31 00:00:00.0',
          },
          prestige_level: '0',
          online_status: '0',
          outfit_member: {
            outfit_id: '37512545478660293',
            outfit: {
              leader_character_id: '5428011263355078529',
              leader: { faction_id: '1' },
            },
          },
        },
        {
          character_id: '5429210842260871009',
          name: { first: 'Number5Johnny', first_lower: 'number5johnny' },
          faction_id: '4',
          head_id: '5',
          title_id: '0',
          times: {
            creation: '1641022496',
            creation_date: '2022-01-01 07:34:56.0',
            last_save: '1675632997',
            last_save_date: '2023-02-05 21:36:37.0',
            last_login: '1675612529',
            last_login_date: '2023-02-05 15:55:29.0',
            login_count: '229',
            minutes_played: '36020',
          },
          certs: {
            earned_points: '311191',
            gifted_points: '11132',
            spent_points: '300806',
            available_points: '21517',
            percent_to_next: '0.1351111111109',
          },
          battle_rank: { percent_to_next: '54', value: '61' },
          profile_id: '190',
          daily_ribbon: {
            count: '0',
            time: '1675555200',
            date: '2023-02-05 00:00:00.0',
          },
          prestige_level: '3',
          online_status: '0',
          outfit_member: {
            outfit_id: '37588148218236901',
            outfit: {
              leader_character_id: '5429269171559319377',
              leader: { faction_id: '3' },
            },
          },
        },
      ],
      returned: 2,
    })
    const reply = await runCommand('+isbruonline')
    expect(requestData).toMatchSnapshot()
    expect(reply).toEqual('No, Bru is offline.')
  })

  it('should reverse the reply when isbruoffline alias is used', async () => {
    const requestData = mockCensusApi({
      character_list: [
        {
          character_id: '5428631729584582401',
          name: { first: 'MiroitoVS', first_lower: 'miroitovs' },
          faction_id: '1',
          head_id: '5',
          title_id: '0',
          times: {
            creation: '1502020291',
            creation_date: '2017-08-06 11:51:31.0',
            last_save: '1659281631',
            last_save_date: '2022-07-31 15:33:51.0',
            last_login: '1659269547',
            last_login_date: '2022-07-31 12:12:27.0',
            login_count: '305',
            minutes_played: '20631',
          },
          certs: {
            earned_points: '63245',
            gifted_points: '8742',
            spent_points: '62563',
            available_points: '9424',
            percent_to_next: '0.888',
          },
          battle_rank: { percent_to_next: '35', value: '88' },
          profile_id: '20',
          daily_ribbon: {
            count: '0',
            time: '1659225600',
            date: '2022-07-31 00:00:00.0',
          },
          prestige_level: '0',
          online_status: '10',
          outfit_member: {
            outfit_id: '37512545478660293',
            outfit: {
              leader_character_id: '5428011263355078529',
              leader: { faction_id: '1' },
            },
          },
        },
        {
          character_id: '5429210842260871009',
          name: { first: 'Number5Johnny', first_lower: 'number5johnny' },
          faction_id: '4',
          head_id: '5',
          title_id: '0',
          times: {
            creation: '1641022496',
            creation_date: '2022-01-01 07:34:56.0',
            last_save: '1675632997',
            last_save_date: '2023-02-05 21:36:37.0',
            last_login: '1675612529',
            last_login_date: '2023-02-05 15:55:29.0',
            login_count: '229',
            minutes_played: '36020',
          },
          certs: {
            earned_points: '311191',
            gifted_points: '11132',
            spent_points: '300806',
            available_points: '21517',
            percent_to_next: '0.1351111111109',
          },
          battle_rank: { percent_to_next: '54', value: '61' },
          profile_id: '190',
          daily_ribbon: {
            count: '0',
            time: '1675555200',
            date: '2023-02-05 00:00:00.0',
          },
          prestige_level: '3',
          online_status: '10',
          outfit_member: {
            outfit_id: '37588148218236901',
            outfit: {
              leader_character_id: '5429269171559319377',
              leader: { faction_id: '3' },
            },
          },
        },
      ],
      returned: 2,
    })
    const reply = await runCommand('+isbruoffline')
    expect(requestData).toMatchSnapshot()
    expect(reply).toEqual(
      'No, Bru is online as **MiroitoVS** VS and **Number5Johnny** NS TR.',
    )
  })
})
