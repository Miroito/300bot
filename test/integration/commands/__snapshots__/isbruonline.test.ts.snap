// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isbruonline should reverse the reply when isbruoffline alias is used 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character",
  "searchParams": Object {
    "c:join": "outfit_member^show:outfit_id^inject_at:outfit_member(outfit^inject_at:outfit^show:leader_character_id(character^on:leader_character_id^to:character_id^show:faction_id^inject_at:leader))",
    "c:resolve": "online_status",
    "character_id": Array [
      "miroitovs",
      "number5johnny",
    ],
  },
}
`;

exports[`isbruonline should say Bru is offline 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character",
  "searchParams": Object {
    "c:join": "outfit_member^show:outfit_id^inject_at:outfit_member(outfit^inject_at:outfit^show:leader_character_id(character^on:leader_character_id^to:character_id^show:faction_id^inject_at:leader))",
    "c:resolve": "online_status",
    "character_id": Array [
      "miroitovs",
      "number5johnny",
    ],
  },
}
`;

exports[`isbruonline should say Bru is online and list the character(s) he is online with 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/character",
  "searchParams": Object {
    "c:join": "outfit_member^show:outfit_id^inject_at:outfit_member(outfit^inject_at:outfit^show:leader_character_id(character^on:leader_character_id^to:character_id^show:faction_id^inject_at:leader))",
    "c:resolve": "online_status",
    "character_id": Array [
      "miroitovs",
      "number5johnny",
    ],
  },
}
`;
